### PR TITLE
feat: bunch of updates to get google login working again

### PR DESCRIPTION
**Pull request by markusahlstrand**
Removed the old auth-provider and switch to use connections (as I forgot I had this before..). Now the connections are dynamically loaded so no more hard-coding.

Changed to use the id-token from google to get the profile, so we don't need to hit the profile endpoint.

### CODE REVIEW

File: src/authentication-flows/social.ts [hunk 1]
```diff
@@ -20,8 +21,10 @@ export async function socialAuth(
   connection: string,
   authParams: AuthParams
 ) {
-  const oauthProvider = client.authProviders.find((p) => p.name === connection);
```
Comment (markusahlstrand) [inline src/authentication-flows/social.ts:23]: Rename from authProvider to connection.. More in line with the naming of endpoints.

File: src/authentication-flows/social.ts [hunk 2]
```diff
@@ -62,32 +65,40 @@ export async function socialAuthCallback({
...
 
   const token = await oauth2Client.exchangeCodeForToken(code);
 
-  const oauth2Profile = await oauth2Client.getUserProfile(token.access_token);
+  const oauth2Profile = parseJwt(token.id_token!);
 
   const userId = getId(client.tenantId, oauth2Profile.email);
   const user = env.userFactory.getInstanceByName(userId);
 
+  console.log(
```
Comment (markusahlstrand) [inline src/authentication-flows/social.ts:89]: Oops. left-overs. Will remove.

File: src/hooks/update-client.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
+import { Env, Client } from "../types";
```
Comment (markusahlstrand) [inline src/hooks/update-client.ts:1]: Moved this to a separate file. Was part of the application before, but.. we need to update this snapshot of the clients in kv-storage if any of the related entities are updated.

File: src/migrate/migrations/2022-12-11T09:17:35_init.ts [hunk 4]
```diff
@@ -42,21 +42,6 @@ export async function up(db: Kysely<any>): Promise<void> {
     .addColumn("modified_at", "varchar")
     .execute();
 
-  await db.schema
```
Comment (markusahlstrand) [inline src/migrate/migrations/2022-12-11T09:17:35_init.ts:45]: Remove the old auth-providers.

File: src/models/User.ts [hunk 5]
```diff
@@ -238,7 +238,9 @@ export const userRouter = router({
           .array(
             z.object({
               name: z.string(),
-              profile: z.record(z.union([z.string(), z.boolean()])).optional(),
+              profile: z
+                .record(z.union([z.string(), z.boolean(), z.number()]))
```
Comment (markusahlstrand) [inline src/models/User.ts:242]: We store the iat and exp that are numbers so we need to allow numbers here as well.

File: src/routes/tsoa/applications.ts [hunk 6]
```diff
@@ -113,6 +62,27 @@ export class ApplicationsController extends Controller {
     return application;
   }
 
+  @Delete("{id}")
+  @Security("oauth2", [])
+  public async deleteConnection(
```
Comment (markusahlstrand) [inline src/routes/tsoa/applications.ts:67]: Added a bunch of endpoints to be able to manage the entities from the admin.

File: src/utils/jwt.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import { decode } from "./base64";
+
+export function parseJwt(token: string): any {
+  const base64Url = token.split(".")[1];
+  const base64 = base64Url.replace(/-/g, "+").replace(/_/g, "/");
+  console.log("Base64 payload: " + base64);
+
+  const jsonPayload = decode(base64);
+
+  return JSON.parse(jsonPayload);
```
Comment (danielgent) [inline src/utils/jwt.ts:10]: Is there no library to do this? :thinking: 

Is this because we're not on Node *or* in a browser?

I feel like you should have unit tests for something so crucial :thinking: 

But it seems like something I'd use an NPM package to do to handle edge cases

*also* there are lots of console logs but if you want that then :+1:

File: src/services/oauth2-client.ts [hunk 8]
```diff
@@ -1,5 +1,6 @@
 export interface TokenResponse {
   access_token: string;
+  id_token?: string;
```
Comment (markusahlstrand) [inline src/services/oauth2-client.ts:3]: Yes, we want the id-token..

Comment (danielgent) [inline src/services/oauth2-client.ts:3]: and then the token-service will read this token and make its own token, much like we're doing now with Auth0's id token :+1:

File: src/utils/jwt.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { decode } from "./base64";
+
+export function parseJwt(token: string): any {
```
Comment (markusahlstrand) [inline src/utils/jwt.ts:3]: Tiny wrapper to pull the payload from a jwt. We should probably do a bit of checks here to make sure the at_hash and other fun stuff matches the access token.

File: src/utils/jwt.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+  const base64Url = token.split(".")[1];
+  const base64 = base64Url.replace(/-/g, "+").replace(/_/g, "/");
+  console.log("Base64 payload: " + base64);
```
Comment (markusahlstrand) [inline src/utils/jwt.ts:6]: Oops. another console.log.

File: src/utils/jwt.ts (hunk 7 above)
Comment (markusahlstrand) [reply src/utils/jwt.ts]: It's a base64 decode, so not really worth a package for it. We have a bit more fancy code that we're running for validating the jwts in other projects, but so far it's not really needed.

File: src/authentication-flows/social.ts (hunk 1 above)
Comment (danielgent) [reply src/authentication-flows/social.ts:23]: If we were the British Museum we'd have an Auth0 section :smile:

File: src/authentication-flows/social.ts [hunk 11]
```diff
@@ -62,32 +65,31 @@ export async function socialAuthCallback({
...
   }
 
   const oauth2Client = env.oauth2ClientFactory.create(
-    oauthProvider,
+    connection,
     `${client.loginBaseUrl}callback`,
     state.authParams.scope?.split(" ") || []
   );
 
   const token = await oauth2Client.exchangeCodeForToken(code);
```
Comment (danielgent) [inline src/authentication-flows/social.ts:82]: this var isn't actually a token though... what to call it though... :shrug:

Comment (danielgent) [reply src/authentication-flows/social.ts:82]: auth0 would call it `exchangeCodeForTokenResponse` IIRC

File: src/migrate/migrations/2022-12-11T09:17:35_init.ts (hunk 4 above)
Comment (danielgent) [reply src/migrate/migrations/2022-12-11T09:17:35_init.ts:45]: ok. So you've removed their creation from the migration script, and then you've manually removed their current instances from cloudflare?

I'm fine with that.  Seems excessive to remove them in another migration script :smile:

File: src/models/User.ts (hunk 5 above)
Comment (danielgent) [reply src/models/User.ts:242]: TIL zod record

**Review by danielgent — APPROVED** [review]
